File: backend/app/validation.py
"""Server-side re-validation of handwriting samples.

Mirrors the client-side checks in frontend/src/lib/validation.ts. The client
is the primary gate (it has the fast feedback loop), but the server must not
trust it: a malformed or bypassed client could otherwise poison the dataset.
"""

from app.config import get_settings
from app.models import SampleUpload

MULTI_ROW_GAP_RATIO = 0.35  # fraction of canvas height treated as a "row gap"


class SampleValidationError(Exception):
    def __init__(self, message: str):
        self.message = message
        super().__init__(message)


def validate_sample(sample: SampleUpload) -> None:
    settings = get_settings()

    if sample.stroke_count <= 0 or not sample.strokes:
        raise SampleValidationError("Canvas is empty — no strokes recorded.")

    if sample.point_count < settings.min_points_per_sample:
        raise SampleValidationError(
            f"Too few points recorded ({sample.point_count} < {settings.min_points_per_sample})."
        )

    for stroke in sample.strokes:
        for point in stroke:
            if not (0 <= point.x <= sample.canvas_width) or not (0 <= point.y <= sample.canvas_height):
                raise SampleValidationError("Stroke point falls outside canvas bounds.")

    bbox = sample.bounding_box
    if bbox.maxX < bbox.minX or bbox.maxY < bbox.minY:
        raise SampleValidationError("Invalid bounding box for recorded strokes.")


def occupies_full_height(sample: SampleUpload) -> bool:
    bbox = sample.bounding_box
    height_used = bbox.maxY - bbox.minY
    return height_used >= sample.canvas_height * 0.85


def looks_multi_row(sample: SampleUpload) -> bool:
    """Heuristic: cluster stroke vertical centers; flag if they split into
    two or more bands separated by a gap wider than MULTI_ROW_GAP_RATIO of
    the canvas height, which suggests writing wrapped onto a second line."""
    if len(sample.strokes) < 2:
        return False

    centers = sorted(
        sum(p.y for p in stroke) / len(stroke) for stroke in sample.strokes if stroke
    )
    if len(centers) < 2:
        return False

    gap_threshold = sample.canvas_height * MULTI_ROW_GAP_RATIO
    return any(b - a > gap_threshold for a, b in zip(centers, centers[1:]))

File: frontend/src/lib/validation.ts
import {
  CANVAS_HEIGHT,
  CANVAS_WIDTH,
  FULL_HEIGHT_WARNING_RATIO,
  MIN_POINTS_PER_SAMPLE,
  MULTI_ROW_GAP_RATIO,
} from "./constants";
import type { BoundingBox, Stroke } from "./types";

export interface ValidationResult {
  errors: string[];
  warnings: string[];
}

export function computeBoundingBox(strokes: Stroke[]): BoundingBox {
  let minX = Infinity;
  let minY = Infinity;
  let maxX = -Infinity;
  let maxY = -Infinity;

  for (const stroke of strokes) {
    for (const point of stroke) {
      if (point.x < minX) minX = point.x;
      if (point.y < minY) minY = point.y;
      if (point.x > maxX) maxX = point.x;
      if (point.y > maxY) maxY = point.y;
    }
  }

  if (!Number.isFinite(minX)) {
    return { minX: 0, minY: 0, maxX: 0, maxY: 0 };
  }

  return { minX, minY, maxX, maxY };
}

export function countPoints(strokes: Stroke[]): number {
  return strokes.reduce((sum, stroke) => sum + stroke.length, 0);
}

export function validateSample(
  strokes: Stroke[],
  canvasWidth: number = CANVAS_WIDTH,
  canvasHeight: number = CANVAS_HEIGHT,
): ValidationResult {
  const errors: string[] = [];
  const warnings: string[] = [];

  const pointCount = countPoints(strokes);

  if (strokes.length === 0 || pointCount === 0) {
    errors.push("Canvas is empty. Please write the sentence before continuing.");
    return { errors, warnings };
  }

  if (pointCount < MIN_POINTS_PER_SAMPLE) {
    errors.push("Too few points were recorded. Please write more carefully with the stylus.");
  }

  const outOfBounds = strokes.some((stroke) =>
    stroke.some((point) => point.x < 0 || point.x > canvasWidth || point.y < 0 || point.y > canvasHeight),
  );
  if (outOfBounds) {
    errors.push("Some strokes were drawn outside the canvas area.");
  }

  if (errors.length > 0) {
    return { errors, warnings };
  }

  const bbox = computeBoundingBox(strokes);
  const heightUsed = bbox.maxY - bbox.minY;
  if (heightUsed >= canvasHeight * FULL_HEIGHT_WARNING_RATIO) {
    warnings.push("Your writing occupies nearly the entire canvas height. Please try to write on a single line.");
  }

  if (looksMultiRow(strokes, canvasHeight)) {
    warnings.push("Your writing may span multiple rows. Please make sure it's a single line.");
  }

  return { errors, warnings };
}

export function looksMultiRow(strokes: Stroke[], canvasHeight: number = CANVAS_HEIGHT): boolean {
  if (strokes.length < 2) return false;

  const centers = strokes
    .filter((stroke) => stroke.length > 0)
    .map((stroke) => stroke.reduce((sum, p) => sum + p.y, 0) / stroke.length)
    .sort((a, b) => a - b);

  if (centers.length < 2) return false;

  const gapThreshold = canvasHeight * MULTI_ROW_GAP_RATIO;
  for (let i = 1; i < centers.length; i += 1) {
    if (centers[i] - centers[i - 1] > gapThreshold) {
      return true;
    }
  }
  return false;
}
